### DISCUSSION TITLE
Ship Map Creator

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#38] [Ideas]
This discussion is much like https://github.com/Thorium-Sim/thorium-nova/discussions/34, except instead of a creator for the universe, we need a creator for the ship map.

First, a few definitions. The ship map serves 3 purposes:

- A way to show the decks and rooms on the ship
- A way to see crew members and cargo containers as they move around the ship (think the Marauders' Map from Harry Potter)
- A way to view utilities on the ship, like power conduits, water pipes, networking equipment, etc.

To begin with, the creator needs to be able to define the following properties of a ship.
- What decks the ship has
- What rooms each deck has
- Corridors around the ship where the crew can walk around
- A way to visually display these decks and rooms
- A definition for where crew members are able to walk so they don't walk through walls.

**Comment by alexanderson1993** [comment]
Decks and rooms are defines separately from the visualization aspect. Decks should be nameable, with "Deck 1, Deck 2, ..." being the default, but with the option to do "A Deck, B Deck, ..." etc. (Lets be honest, I want to recreate the NSEA Protector in Thorium Nova someday 😝).

For decks and rooms visualization, SVGs make a lot of sense for this. They are relatively small, and you can add metadata to the SVG elements to identify each with a deck and room. Since we're doing this as a web app, you can easily render the SVG and add interactivity to click on the decks and rooms.

The idea is that each deck would be its own SVG, or its own layer in an SVG, and each room is a polygon shape in the SVG.

As for defining the movement of crew members, we probably want to use a graph structure of some kind, with nodes and edges. Nodes could exist inside hallways, or they could be associated with specific rooms. Edges connect all the places where a node connects. Crew members can move along edges between nodes. This also makes pathfinding really easy using Dijkstra's or A* algorithms. 

For movement from deck to deck, a node could specify an edge that connects to a node on a separate deck. These would be placed in elevator/turbolifts or maintenance ladders. Maybe edges could have weights to discourage regular crew members from going along paths reserved for maintenance crew members.

Nodes that exist in larger areas could have a property that defines a radius. That way, when a crew member's dot on the Marauder's map enters a room, they don't just sit in the middle - the radius defines an area where they can randomly wander while they are "in" that node, which adds to the realism just a little bit.

I think that covers it.

**Reply by Xansta** [reply]
Ship maps are an area that is sorely lacking in Artemis and Empty Epsilon. In Artemis, only the Engineer ever sees something resembling decks and corridors and that is only represented by a 3D array of nodes connected by lines. A damage control team is a diamond that moves along the lines to various nodes. The nodes are textually identified as hallways and various systems that might need repair after taking damage. For Empty Epsilon, there is only a 2D series of rectangles and squares connected by doors. The various ship systems reside in the rectangle and squares. There are different numbers of repair crew that can be directed to different rooms to work on the systems. They are represented by an overhead view of a walking person.

For both Artemis and EE, there are no conduits for wiring or flow (coolant, water, air, sewage, etc.). There are no tubes for maintenance crew to crawl around in. There are no medical bays, crew quarter, auxiliary bridges, weapons lockers, cargo bays, etc. I always thought it would add to missions if these things were expressed graphically or had some bearing on game play. I wrote in a cargo system via LUA for EE, but it was essentially verbal. You could get different types of cargo up to a particular limit and you could use the cargo for mission goals.

I could imagine a scenario where instead of the weapons officer clicking a button that loads the torpedo, a dedicated human participant is designated to either move the torpedo from the weapons storage rack into the tube loading mechanism or be tasked with changing the prepared weapons load out in anticipation of particular types of combat. This kind of thing requires more forethought and communication among the crew to accomplish the mission. Admittedly, it reduces the video game aspect, too.

When I looked at Thorium Classic, I found it difficult to figure out how to make a ship. I want designers to have at their fingertips the ability to create the ship they've always imagined, but I also want there to be a way to get a ship built quickly. Let's face it, we're talking about people with imagination, so even if the ship map shows several iterations of squares with text labels that say "Bridge," "Engine Room," and "Cargo Bay," etc. many people are going to be able to take that into their brain and run with it. So, I think there should be layers of creation complexity. The simplest of tools should be at the forefront for a quick construction of a ship. Simple shapes for rooms, automated or default creation of conduits and hallways, maybe a set of three or four basic ships of varying sizes on which to build customized versions. The more complex tools should be there, maybe behind a menu selection, for those that want to spend more time on the design and creation of their ships.

**Reply by alexanderson1993** [reply]
Yes to most everything you see. 

The way I see it, internal stuff on the ship for security, damage control/maintenance, cargo, and medical are what make it possible to go from 6 to 12 crew. 

Here's an abbreviated list of internal stuff in Thorium Classic:

- Security Decks
- Security Scans
- Internal Communications
- Damage Control Teams
- Security Teams
- Medical Teams
- Cargo Control
- Coolant Control
- Power Distribution
- Reactor Control
- Torpedo Loading
- Crew Roster
- Security Armory
- Damage Control Armory
- Medical Armory
- Computer Core
- Decontamination
- Medical Roster
- Sickbay
- Communications Review (random crew members sending out messages that have to be reviewed, in case one of them is a spy or sending secret info)

Obviously,  many of these also affect stuff that happens outside of the ship, but there's a lot of interplay there. We can put these into three categories:

- People (Teams, Sickbay, Communications)
- Cargo (Cargo & Coolant, Armories, Reloading torpedos and other consumables)
- Utilities (Reactor, Power distribution)

There's so much potential to add other things to this that just haven't been dreamed up yet. Having individual crew members be agents that can do stuff like send random messages that have to be reviewed by the bridge, and having one of them be a spy... That's really interesting.

> When I looked at Thorium Classic, I found it difficult to figure out how to make a ship.

I imagine this will still be the case for many things in Thorium Nova. It will be a complicated piece of software that will take a good bit of effort to learn how to use. Making the configuration as easy as possible, having intuitive user interfaces, and providing in-app docs and training will be a big help.

> Let's face it, we're talking about people with imagination.

True true. Unfortunately, I have no imagination, so I need a visualization of what you are talking about to get behind it.

What would be _sick_ and also really hard is if you could take a deck and room listing and a 3D ship model and the system was able to procedurally generate SVGs for the rooms on each deck, including creating the walkable path node graph. Don't like the way the system laid it out? Randomize the seed. Plus, this makes storing the SVGs really easy - you don't! You just store the random seed and it generates the deck maps on the fly or if generating the deck maps is an expensive operation, generate them when the flight starts.

This would be an _awesome_ project if someone wanted to take it on, and would be the holy grail for this sort of thing if done right. Really easy to use, really good user experience.

**Reply by Xansta** [reply]
One of the things I really like about Empty Epsilon is that it comes with built in tutorials. I think we should take that approach with Nova. Specifically, detailed tutorials on ship making, scenario/mission/timeline writing, walk-throughs on cards and any sets of cards we come up with (stations, tabbed cards, etc.), how to be a flight director/GM (examples of any controls available), etc.

**Reply by alexanderson1993** [reply]
Way ahead of you ;) https://github.com/Thorium-Sim/thorium-nova/issues/39

The idea is these docs can be linked to as their own pages, or can be pulled up anywhere in the app as a slide-over panel or something similar. Imagine working on a mission, clicking the 'Docs' button, typing in a little search query to see how to configure a timeline input, and finding exactly what you're looking for, all without leaving the page you were on.

So yes, this will be a thing.

**Reply by Xansta** [reply]
There are text based descriptions for how to write an EE scenario that includes actual code samples. It also points to the existing scenarios as potential examples. It takes you through four stages of scenario script development. That's good as far as it goes, but the tutorials go a step further. They take you through examples of working the various controls of the ship by making you perform those actions in the tutorial.

I don't think that that kind of tutorial is possible for every aspect of Nova, but that's like the gold standard I was aspiring to. It should be possible to create such hands on tutorials for the operation of the cards. The challenge to the participants should not be learning how to operate the controls but how to use the controls with your fellow crew to complete the mission.

For the complicated things like ship creation, there should be a couple of levels of documentation. The reference material and something that walks you through simple tasks followed by more complex examples. It's possible that the reference material can serve both of those purposes. However, what I've seen happen on other projects is that a feature will be added and the developer of that feature knows how to use it and apply it, but is not able to describe very well that feature to other potential users. The reference material is accurate, but not sufficient for non-developers or less experienced users.

**Reply by alexanderson1993** [reply]
> but the tutorials go a step further.

The way you describe them, they sound like videos. Is that accurate? I do hope to have multiple levels of training for the stations themselves. The fact that we decided stations can't be reconfigured means its easier to make video and audio training, since we know which cards will be present.

I would love to talk through more of the different levels of documentation you have in mind. For the crew, I was thinking of having guides for each of the stations as well as some tips and tricks or novel ideas for how they can use the controls. I don't think I can guess how the Flight Director docs will be until we start fleshing out how the FD controls will work.

**Reply by astrolamb-gaming** [reply]
> What would be sick and also really hard is if you could take a deck and room listing and a 3D ship model and the system was able to procedurally generate SVGs for the rooms on each deck, including creating the walkable path node graph. Don't like the way the system laid it out? Randomize the seed. Plus, this makes storing the SVGs really easy - you don't! You just store the random seed and it generates the deck maps on the fly or if generating the deck maps is an expensive operation, generate them when the flight starts.

This sounds a bit more in line with the programming I'm used to, with the caveat that I have very little experience with SVG. But I can give it a shot at some point.

**Reply by alexanderson1993** [reply]
Oooooh.... This is exactly what I want to hear!

I wouldn't worry about the SVGs too much. The API isn't terribly complicated, and to begin with you can just store the room locations as `{x, y, width, height}`. It would be amazing if you could figure out how to make the rooms follow the curvature of the ship and corridors, but that's not strictly necessary to begin with.

**Comment by Xansta** [comment]
It's not a video. More like, "set your impulse engine to 100%." Then the tutorial waits for you to do that. Afterwards, "set your course to 200 degrees," etc. It's literally hands on. You should try it out. Time: download + unzip + 2 minutes.

**Reply by alexanderson1993** [reply]
> Time: download + unzip + 2 minutes.

I'm on a Mac, so this includes compilation 😁 

It's been a while since I played around with EE, so I'll give that a shot today.

**Comment by Xansta** [comment]
I've been running it under Wine on my Mac. Recent changes to EE combined with my older OS (10.9.5) have made compilation challenging for me. I compile on my Linux boxes. I should have remembered that you use Mac

**Reply by alexanderson1993** [reply]
No worries. I got it up and running just fine this morning, and I saw what you were talking about. I'm hoping to do something similar to that with the on-screen text training - similar to what Thorium Classic has, but that is able to respond to them actually using the controls.
